src/core/Future.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * This file is part of Swoole.
+ *
+ * @link     https://www.swoole.com
+ * @contact  team@swoole.com
+ * @license  https://github.com/swoole/library/blob/master/LICENSE
+ */
+
+declare(strict_types=1);
+
+namespace Swoole;
+
+use Swoole\Coroutine\Channel;
+use Throwable;
+
+class Future
+{
+    /**
+     * @var callable
+     */
+    private $func;
+
+    public function __construct(callable $func)
+    {
+        $this->func = $func;
+    }
+
+    public static function create(callable $func): self
+    {
+        return new self($func);
+    }
+
+    public static function join(array $futures, float $timeout = -1): array
```

Comment:
Your implementation is good and readable. But, I would put some minor tweaks (not functional ones, just a different style and one less variable to decrement) into it:

```
   public static function join(array $futures, float $timeout = -1): array
    {
        $ch = new Channel(count($futures));

        Coroutine::join(array_map(static function (Future $future) use ($ch): int {
            return $future->run($ch);
        }, $futures));

        $rets = [];
        while ($ret = $ch->pop($timeout)) {
            if ($ret instanceof Throwable) {
                throw $ret;
            }

            $rets[] = $ret;
        }

        return $rets;
    }
```